test/client_certificate_test.dart

```diff
@@ -0,0 +1,131 @@
+@TestOn('vm')
+import 'dart:async';
+import 'dart:io';
+
+import 'package:grpc/grpc.dart';
+import 'package:test/test.dart';
+
+import 'src/generated/echo.pbgrpc.dart';
+
+class EchoService extends EchoServiceBase {
+  @override
+  Future<EchoResponse> echo(ServiceCall call, EchoRequest request) async {
+    final subject = call.clientCertificate?.subject;
+    return (EchoResponse()..message = subject ?? 'NO CERT');
+  }
+
+  @override
+  Stream<ServerStreamingEchoResponse> serverStreamingEcho(
+      ServiceCall call, ServerStreamingEchoRequest request) {
+    // TODO: implement serverStreamingEcho
+    throw UnimplementedError();
+  }
+}
+
+const String address = 'localhost';
+const int port = 19360;
+Future<void> main() async {
+  test('Client certificate required', () async {
+    //Server
+    final server = Server(
+      [EchoService()],
+    );
+    final serverContext =
+        SecurityContextChannelCredentials.baseSecurityContext();
+
+    serverContext.useCertificateChain('test/data/localhost.crt');
+    serverContext.usePrivateKey('test/data/localhost.key');
+    serverContext.setTrustedCertificates('test/data/localhost.crt');
+    final ServerCredentials serverCredentials =
+        SecurityContextServerCredentials(serverContext);
+
+    await server.serve(
+        address: address,
+        port: port,
+        security: serverCredentials,
+        requireClientCertificate: true);
+
+    //Client
+    final channelContext =
+        SecurityContextChannelCredentials.baseSecurityContext();
+    channelContext.useCertificateChain(
+      'test/data/localhost.crt',
+    );
+    channelContext.usePrivateKey('test/data/localhost.key');
+    final channelCredentials = SecurityContextChannelCredentials(channelContext,
+        onBadCertificate: (cert, s) {
+      return true;
+    });
+    final channel = ClientChannel(address,
+        port: port, options: ChannelOptions(credentials: channelCredentials));
+    final client = EchoServiceClient(channel);
+    expect((await client.echo(EchoRequest())).message, '/CN=localhost');
+    await channel.shutdown();
+    await server.shutdown();
+  });
+
+  test('Client certificate not required', () async {
+    //Server
+    final server = Server(
```

Comment:
Consider moving `server` setup into a helper method because it is duplicated between two tests.